core/utils/utils.go

```diff
@@ -866,65 +867,53 @@ const (
 
 // StartOnce sets the state to Started
 func (once *StartStopOnce) StartOnce(name string, fn func() error) error {
-	once.Lock()
-	defer once.Unlock()
+	success := once.state.CAS(int32(StartStopOnce_Unstarted), int32(StartStopOnce_Started))
 
-	if once.state != StartStopOnce_Unstarted {
+	if !success {
 		return errors.Errorf("%v has already started once", name)
 	}
-	once.state = StartStopOnce_Started
+
+	once.Lock()
+	defer once.Unlock()
 
 	return fn()
 }
 
 // StopOnce sets the state to Stopped
 func (once *StartStopOnce) StopOnce(name string, fn func() error) error {
-	once.Lock()
-	defer once.Unlock()
+	success := once.state.CAS(int32(StartStopOnce_Started), int32(StartStopOnce_Stopped))
 
-	if once.state != StartStopOnce_Started {
-		return errors.Errorf("%v has already stopped once", name)
+	if !success {
+		return errors.Errorf("%v has already started once", name)
 	}
-	once.state = StartStopOnce_Stopped
+
+	once.Lock()
+	defer once.Unlock()
 
 	return fn()
 }
 
 // OkayToStart checks if the state may be started
```

Comment:
minor, but the comment seems outdated, as it not only checks, but also changes the state. Similar for OkayToStop